File: ios/Packages/AppFeatures/Sources/AwradFeature/FixedWirdSlots.swift
import Foundation

/// The four awrad offered as a one-tap starting board (client decision,
/// 2026-08-12, reversing the 2026-08-09 "seeded for everyone, forever"
/// decision): قيام الليل، ورد يومي من القرآن، أذكار الصباح، أذكار المساء. Nothing
/// is seeded automatically — a user adds them via "أضف ورد اليوم" — and once on
/// the board they are ordinary wirds: tickable, retargetable, and deletable
/// like any wird the user creates themselves.
///
/// The ids are stable literals rather than UUIDs precisely because they have to
/// be recognisable across launches and across the two platforms — seeding is
/// "add the ids that are missing", which is what makes it idempotent.
public enum FixedWirdSlot: String, CaseIterable, Sendable {
    case qiyamAlLayl
    case dailyQuran
    case morningAzkar
    case eveningAzkar

    /// Stable, human-readable, identical on iOS and Android.
    public var wirdId: String {
        switch self {
        case .qiyamAlLayl: "fixed-qiyam-al-layl"
        case .dailyQuran: "fixed-daily-quran"
        case .morningAzkar: "fixed-morning-azkar"
        case .eveningAzkar: "fixed-evening-azkar"
        }
    }

    /// Localization key the app layer resolves. The *resolved* string is what
    /// gets persisted into `Wird.name` at seeding time — same as a template-created
    /// wird, whose name is likewise frozen in the locale it was created in.
    public var nameKey: String {
        switch self {
        case .qiyamAlLayl: "awrad.fixed.qiyam_al_layl"
        case .dailyQuran: "awrad.fixed.daily_quran"
        case .morningAzkar: "awrad.fixed.morning_azkar"
        case .eveningAzkar: "awrad.fixed.evening_azkar"
        }
    }

    /// Fallback used by tests and by any caller that has no string resolver.
    public var defaultName: String {
        switch self {
        case .qiyamAlLayl: "قيام الليل"
        case .dailyQuran: "ورد يومي من القرآن"
        case .morningAzkar: "أذكار الصباح"
        case .eveningAzkar: "أذكار المساء"
        }
    }

    /// Vocabulary matched to the backend `wird_templates` seed so the stats
    /// aggregation keeps working: `unit == "pages"` feeds the Qur'an-pages tile.
    public var type: String {
        switch self {
        case .qiyamAlLayl: "qiyam"
        case .dailyQuran: "quran_reading"
        case .morningAzkar, .eveningAzkar: "azkar"
        }
    }

    public var unit: String {
        switch self {
        case .qiyamAlLayl: "rakaat"
        case .dailyQuran: "pages"
        case .morningAzkar, .eveningAzkar: "times"
        }
    }

    /// Deliberately small. A fixed slot the user never asked for must not be the
    /// thing that makes the day uncompletable; anyone who wants more retargets it.
    public var defaultTarget: Int {
        switch self {
        case .qiyamAlLayl: 2
        case .dailyQuran: 1
        case .morningAzkar, .eveningAzkar: 1
        }
    }

    public var frequency: String { "daily" }

    /// Hour this slot's daily reminder fires at, overriding the user's single
    /// reminder time. `nil` means "use the user's time".
    ///
    /// Without this, four always-present wirds would fire four notifications in
    /// the same minute — and a "did you do أذكار الصباح?" nudge at 20:00 is asking
    /// about a window that closed hours ago. Each slot is asked about when it is
    /// still actionable; the user's own wirds keep the time they configured.
    public var reminderHour: Int? {
        switch self {
        case .qiyamAlLayl: 22
        case .dailyQuran: nil
        case .morningAzkar: 8
        case .eveningAzkar: 17
        }
    }

    /// The prayer this slot can be anchored to, when offering the
    /// "follow the prayer instead of the clock" option.
    ///
    /// Only the two azkar slots have one. أذكار الصباح and أذكار المساء are tied
    /// to a prayer by definition, and their windows move by over an hour across
    /// the year — a fixed time chosen in summer lands at the wrong end of the
    /// window in winter. قيام الليل is anchored to the last third of the night
    /// rather than a prayer, and the Qur'an wird has no natural anchor at all.
    public var anchorPrayer: String? {
        switch self {
        case .morningAzkar: "fajr"
        case .eveningAzkar: "asr"
        case .qiyamAlLayl, .dailyQuran: nil
        }
    }

    /// Default minutes after the anchor prayer. Enough that the prayer itself is
    /// done before the app asks about the adhkar that follow it.
    public var defaultAnchorOffsetMinutes: Int { 30 }

    /// Board / reminder order.
    public var sortOrder: Int { Self.allCases.firstIndex(of: self) ?? 0 }

    public init?(wirdId: String) {
        guard let match = Self.allCases.first(where: { $0.wirdId == wirdId }) else { return nil }
        self = match
    }
}

public enum FixedWirdSlots {
    /// Resolves a slot's display name. The app injects `NSLocalizedString`; the
    /// default keeps the package testable without a bundle.
    public typealias NameResolver = @Sendable (FixedWirdSlot) -> String

    public static let defaultNames: NameResolver = { $0.defaultName }

    public static func isFixed(wirdId: String) -> Bool { FixedWirdSlot(wirdId: wirdId) != nil }

    public static func wird(for slot: FixedWirdSlot, name: NameResolver = defaultNames, now: Date) -> Wird {
        Wird(
            id: slot.wirdId,
            name: name(slot),
            type: slot.type,
            target: slot.defaultTarget,
            unit: slot.unit,
            frequency: slot.frequency,
            createdAt: now,
            isFixed: true
        )
    }

    /// Refreshes the display name of every fixed-slot record to the
    /// *current* resolver's value. Unlike `applied`, this neither adds a
    /// missing slot nor reactivates an archived one, so it is safe to run
    /// passively on every load — a fixed slot's name is not something a user
    /// ever chose (retargeting is the only edit these slots offer), so unlike
    /// a template wird's frozen name, it has no reason to stay frozen under
    /// whatever language happened to be active the first time it was seeded.
    public static func normalized(_ wirds: [Wird], name: NameResolver = defaultNames) -> [Wird] {
        wirds.map { wird in
            guard let slot = FixedWirdSlot(wirdId: wird.id) else { return wird }
            var updated = wird
            updated.name = name(slot)
            return updated
        }
    }

    /// The seeding rule, invoked once when the user taps "أضف ورد اليوم" — not on
    /// every launch anymore, so both platforms and the tests can still agree on
    /// what one tap does.
    ///
    /// - Appends a fresh instance only for a slot whose id is **entirely
    ///   missing**, in canonical order, so a repeat tap can never duplicate a
    ///   slot or reset the target of one the user already has.
    /// - Un-archives a fixed slot that is present but was previously deleted.
    ///   That is exactly what re-adding it means: the same slot, its history
    ///   under that id intact, active again — not a fresh duplicate record.
    public static func applied(to wirds: [Wird], name: NameResolver = defaultNames, now: Date) -> [Wird] {
        var result = normalized(wirds, name: name).map { wird -> Wird in
            guard wird.isFixed || isFixed(wirdId: wird.id), wird.archivedAt != nil else { return wird }
            var repaired = wird
            repaired.archivedAt = nil
            return repaired
        }
        let present = Set(result.map(\.id))
        for slot in FixedWirdSlot.allCases where !present.contains(slot.wirdId) {
            result.append(wird(for: slot, name: name, now: now))
        }
        return result
    }
}

File: android/feature/awrad/src/main/kotlin/com/fatwabot/feature/awrad/FixedWirdSlots.kt
package com.fatwabot.feature.awrad

/**
 * The four awrad offered as a one-tap starting board (client decision,
 * 2026-08-12, reversing the 2026-08-09 "seeded for everyone, forever"
 * decision). Mirror of iOS `FixedWirdSlot`. Nothing is seeded automatically —
 * a user adds them via "أضف ورد اليوم" — and once on the board they are
 * ordinary wirds: tickable, retargetable, and deletable like any wird the
 * user creates themselves.
 *
 * The ids are stable literals rather than UUIDs precisely because they have to
 * be recognisable across launches and across the two platforms — seeding is
 * "add the ids that are missing", which is what makes it idempotent.
 */
enum class FixedWirdSlot(
    /** Stable, human-readable, identical on iOS and Android. */
    val wirdId: String,
    /** Fallback used by tests and by any caller with no string resolver. */
    val defaultName: String,
    /**
     * Vocabulary matched to the backend `wird_templates` seed so the stats
     * aggregation keeps working: `unit == "pages"` feeds the Qur'an-pages tile.
     */
    val type: String,
    val unit: String,
    /**
     * Deliberately small. A fixed slot the user never asked for must not be the
     * thing that makes the day uncompletable; anyone who wants more retargets it.
     */
    val defaultTarget: Int,
    /**
     * Hour this slot's daily reminder fires at, overriding the user's single
     * reminder time. `null` means "use the user's time".
     *
     * Without this, four always-present wirds would fire four notifications in
     * the same minute — and a "did you do أذكار الصباح?" nudge at 20:00 is asking
     * about a window that closed hours ago. Each slot is asked about when it is
     * still actionable; the user's own wirds keep the time they configured.
     */
    val reminderHour: Int?,
) {
    QIYAM_AL_LAYL("fixed-qiyam-al-layl", "قيام الليل", "qiyam", "rakaat", 2, 22),
    DAILY_QURAN("fixed-daily-quran", "ورد يومي من القرآن", "quran_reading", "pages", 1, null),
    MORNING_AZKAR("fixed-morning-azkar", "أذكار الصباح", "azkar", "times", 1, 8),
    EVENING_AZKAR("fixed-evening-azkar", "أذكار المساء", "azkar", "times", 1, 17),
    ;

    val frequency: String get() = "daily"

    /** Board / reminder order. */
    val sortOrder: Int get() = ordinal

    /**
     * The prayer this slot can be anchored to, when offering the "follow the
     * prayer instead of the clock" option. Mirror of iOS `anchorPrayer`.
     *
     * Only the two azkar slots have one: their windows are defined by a prayer
     * and move by over an hour across the year, so a fixed time chosen in summer
     * lands at the wrong end of the window in winter. قيام الليل follows the last
     * third of the night rather than a prayer, and the Qur'an wird has no
     * natural anchor at all.
     */
    val anchorPrayer: String?
        get() = when (this) {
            MORNING_AZKAR -> "fajr"
            EVENING_AZKAR -> "asr"
            QIYAM_AL_LAYL, DAILY_QURAN -> null
        }

    /** Default minutes after the anchor prayer. */
    val defaultAnchorOffsetMinutes: Int get() = 30

    companion object {
        fun forWirdId(wirdId: String): FixedWirdSlot? = entries.firstOrNull { it.wirdId == wirdId }
    }
}

object FixedWirdSlots {
    /**
     * Resolves a slot's display name. The app injects a `Context.getString`
     * lookup; the default keeps the module testable without resources.
     */
    fun interface NameResolver {
        fun name(slot: FixedWirdSlot): String
    }

    val defaultNames = NameResolver { it.defaultName }

    fun isFixed(wirdId: String): Boolean = FixedWirdSlot.forWirdId(wirdId) != null

    fun wird(
        slot: FixedWirdSlot,
        name: NameResolver = defaultNames,
        nowEpochSeconds: Long,
    ): Wird = Wird(
        id = slot.wirdId,
        name = name.name(slot),
        type = slot.type,
        target = slot.defaultTarget,
        unit = slot.unit,
        frequency = slot.frequency,
        createdAtEpochSeconds = nowEpochSeconds,
        isFixed = true,
    )

    /**
     * Refreshes the display name (and repairs `isFixed`) of every fixed-slot
     * record to the *current* resolver's value. Unlike [applied], this
     * neither adds a missing slot nor reactivates an archived one, so it is
     * safe to run passively on every load — a fixed slot's name is not
     * something a user ever chose (retargeting is the only edit these slots
     * offer), so unlike a template wird's frozen name, it has no reason to
     * stay frozen under whatever language happened to be active the first
     * time it was seeded.
     */
    fun normalized(wirds: List<Wird>, name: NameResolver = defaultNames): List<Wird> =
        wirds.map { wird ->
            val slot = FixedWirdSlot.forWirdId(wird.id) ?: return@map wird
            wird.copy(isFixed = true, name = name.name(slot))
        }

    /**
     * The seeding rule, invoked once when the user taps "أضف ورد اليوم" — not on
     * every launch anymore, so both platforms and the tests can still agree on
     * what one tap does.
     *
     * - Appends a fresh instance only for a slot whose id is **entirely
     *   missing**, in canonical order, so a repeat tap can never duplicate a
     *   slot or reset the target of one the user already has.
     * - Un-archives a fixed slot that is present but was previously deleted.
     *   That is exactly what re-adding it means: the same slot, its history
     *   under that id intact, active again — not a fresh duplicate record.
     */
    fun applied(
        wirds: List<Wird>,
        name: NameResolver = defaultNames,
        nowEpochSeconds: Long,
    ): List<Wird> {
        val result = normalized(wirds, name).map { wird ->
            if (!wird.isFixed && !isFixed(wird.id)) {
                wird
            } else {
                wird.copy(isFixed = true, archivedAtEpochSeconds = null)
            }
        }.toMutableList()
        val present = result.mapTo(mutableSetOf()) { it.id }
        for (slot in FixedWirdSlot.entries) {
            if (slot.wirdId !in present) result += wird(slot, name, nowEpochSeconds)
        }
        return result
    }
}
